Integrated thermal model into update_opds

kjbrooks opened the pull request:
I have added the code necessary for the user to update their OPD with the changes due to thermal variation that would be expected with a slew relative to the sun. I have also added notebooks to explain how to do this with WebbPSF, and I have added my notebook for how I created the FITS table that is used for this. 

The reason this is a work in progress is because I had to write a new function to apply the global Hexikes. There existed `_apply_global_zernikes` however it wasn't clear how to apply this to the Hexike coefficients I had since I have always worked with `hexike_basis_wss` and passed in the OTE aperture. If there is a way to do this and combine `_apply_global_zernikes` with the new `_apply_global_hexikes`, I would love to hear it. 

Additionally, I had a lot of trouble with deciding what to do with regards to `rho` and `theta` in this new function. There seem to be many different ways that this is done in `poppy` and since there is no one function (and it seems like most of what is done in `poppy` does not take normalization to the radius of the aperture into account since the aperture is often created after `rho` and `theta`). If there is a better way to do this, or open an issue to create an unified way of creating `rho` and `theta`, please let me know.

mperrin commented:
Thanks Keira - I will take a close look at this and try to figure out answers to some of your good questions above, but it may take a while (post AAS?). Just left Lucie’s phD defense post celebration here in Marseille; back in Baltimore Thurs but then straight into holiday break... I suspect you’re right that my global Zernikes function doesn’t also do hexikes, and having two functions may be most practical. I’ll take a look.

mperrin commented:
Hi @kjbrooks, just to confirm, I'm going to leave the detailed read and review of this for post-AAS. But I haven't forgotten about it!

kjbrooks commented:
That works!

kjbrooks commented:
Just pinging @mperrin on this one to see how this is going and if there are any questions.

kjbrooks commented:
No longer a WIP!

kjbrooks commented:
Everything has been checked and no more changes need to be added. This is all set to be reviewed.

mperrin commented on webbpsf/tests/test_opds.py:
```diff
@@ -22,3 +29,49 @@ def test_enable_adjustable_ote():
+        give the expected output'''
+    otelm = webbpsf.opds.OTE_Linear_Model_WSS()
+    otelm.thermal_slew(delta_time=1.0*u.day)
+    assert np.max(otelm.opd) == 4.1333852193439e-08
```
This test is a dangerous one, since it is looking for exact floating point equality. Finite floating point precision (~ "rounding errors") can give different behavior on different machines, in terms of the last few decimal places. It's better to write tests like this in a way that checks for closeness (within some tolerance) rather than exact equality in every bit. The `np.isclose` function is good for this. It's also good form to give test assertions an informative error message in case the assertion fails.  
```
assert np.isclose(np.max(otelm.opd), 4.13338e-08), "OPD max does not match expected value after 1 day slew."
```

kjbrooks replied:
Thanks! I will make this change.

mperrin commented on webbpsf/tests/test_opds.py:
```diff
@@ -22,3 +29,49 @@ def test_enable_adjustable_ote():
+    # Assert the coefficents
+        for coeff, truth in zip(coeffs, coeff_truth):
+            assert np.round(coeff, decimals=4) == np.round(truth, decimals=4)
+
```
It would be nice to also add a test case that demonstrates the start_angle and end_angle parameter have the right effect if they're changed - it only seems they are used as their default values here.

kjbrooks replied:
Agreed, I will add one

mperrin commented on webbpsf/tests/test_opds.py:
```diff
@@ -22,3 +29,49 @@ def test_enable_adjustable_ote():
+    delta_time = time
+    # Create the thermal model
+    otelm = webbpsf.opds.OTE_Linear_Model_WSS()
+    otelm.delta_time = delta_time
```
This part might become a little more elegant if you instead did this via 
```
otelm.thermal_slew(delta_time, start_angle, end_angle, scaling)
``` 
That'd be more like how we expect end-users to interact with this code, right? By calling that function, rather than by individually assigning values to attributes of the OTE linear model one at a time. In some sense it's a nicer test if we interact with the code the same way a user would. 

(To be clear, what you're doing here works perfectly well too.)

kjbrooks replied:
This is a good point. I had somehow convinced myself that this wouldn't work, but it does. I agree that they way you have described is more how we expect users to use it. I will change it also because it's simpler code.

mperrin commented on webbpsf/opds.py:
```diff
@@ -2164,3 +2273,119 @@ def test2_OPDbender(filename='OPD_RevV_nircam_132.fits'):
+
+class OteThermalModel(object):
+    """
+    Given a time in units of seconds, minutes, hours, or days.
```
Revise this doc string to start with a one-sentence summary of what this class does.

mperrin commented on webbpsf/opds.py:
```diff
@@ -1737,6 +1771,77 @@ def move_sur(self, sur_file, group=None, verbose=False):
+            self.update_opd(display=display)
+
+
+    def thermal_slew_opd(self, segid):
```
Relatedly, I've just noticed that the `.thermal_model` instance also comes up when you try to tab complete. I think that is also something that we do not expect users to interact with directly, right? In which case we could also consider hiding it behind an initial underscore as well: `._thermal_model`.

kjbrooks replied:
That's correct, users should not be able to interact with `.thermal_model`. I will make that update.

kjbrooks commented on webbpsf/tests/test_opds.py:
```diff
@@ -22,3 +29,49 @@ def test_enable_adjustable_ote():
+                            [1 * u.day, 'A1', COEFFS_A1, None],
+                            [0.0 * u.day, 'SM', [0.0], None],
+                            [0.0 * u.day, 'A1', np.zeros(9), None],
+                            [1.0 * u.day, 'D1', [0.0], None])
```
That is not a typo, that is checking to make sure that the code reacts properly when giving it a segment that doesn't exist (we don't want it to return an empty list - which is what it does if you ask it for a segment that doesn't exist - because that will mess things up).

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -1297,9 +1310,30 @@ def _apply_global_zernikes(self):
+        aperture = self._segment_masks != 0
+        # Get size of mask (1024)
+        npix = np.shape(aperture)[0]
+        ap_radius = constants.JWST_CIRCUMSCRIBED_DIAMETER // self.pixelscale.value / 2
```
Agreed. I will remove it.

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -1737,6 +1771,77 @@ def move_sur(self, sur_file, group=None, verbose=False):
+        """ Update the OPD based on a thermal slew between observations.
+
+        Use a delta slew time, beginning and ending angles of the observatory
+        relative to the sun, and any additonal scaling factor to determine the
```
I have rephrased this to be more clear.

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -2164,3 +2273,119 @@ def test2_OPDbender(filename='OPD_RevV_nircam_132.fits'):
+        return coeffs
+
+
+    def get_coeffs(self, segid):
```
Agreed. I will place the current `get_coeffs` function with `get_segment_coeffs`

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -2164,3 +2273,119 @@ def test2_OPDbender(filename='OPD_RevV_nircam_132.fits'):
+        -----------
+        delta_time: astropy units quantity
+            Given a delta time in units of time. If this quantity is an int or
+            float then units of *minutes* will be assumed.
```
I chose it arbitrarily so I am open to changing it and is easy to do. 

I decided to go with hours, but if you think that something else makes more sense, please do let me know.

mperrin replied:
hours seems fine. I agree it's arbitrary. Thanks.

kjbrooks commented on webbpsf/tests/test_opds.py:
```diff
@@ -22,3 +29,49 @@ def test_enable_adjustable_ote():
 
+
+
+GLOBAL_FOCUS = [-0.018251043541410904e-08]
```
I have added a comment about where these values are from. They are the terms that will be returned for delta_time of 1 day based on the most recent table: otelm/thermal_OPD_fitting_parameters_9H_um.fits. If more information is needed, please let me know.

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -2164,3 +2273,119 @@ def test2_OPDbender(filename='OPD_RevV_nircam_132.fits'):
+        delta_time
+
+    """
+    def __init__(self, delta_time):
```
This is a really good point and sounds a lot slicker. It was simple to implement since very few tweaks were required.

kjbrooks commented:
@mperrin I made the recommended changes. I also:

-  updated the notebook to include the creation of PSFs from the OPDs (which I had not included before) 
- added some extra information in `opd_header` so that the user easily can determine what has been done to the OPD
- Lastly, I realized in updating the notebook that the scaling factor was being computed incorrectly (it was making it such that a smaller change in angle would result in larger WFE) so I fixed that. 

When you have a chance, please look over the updates and let me know if any thing needs to be changed.

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -1798,47 +1809,40 @@ def thermal_slew(self, delta_time, start_angle=-5,end_angle=45,
-
         if not delay_update:
             self.update_opd(display=display)
+            self.opd_header['BUNIT'] = 'meter'
```
That makes sense. I will change this.

kjbrooks commented on webbpsf/opds.py:
```diff
@@ -1798,47 +1809,40 @@ def thermal_slew(self, delta_time, start_angle=-5,end_angle=45,
+        delta_time = convert_quantity(delta_time, to_units=u.day) #this returns astropy units quantity
+        self.delta_time = delta_time.value
         self.start_angle = start_angle
         self.end_angle = end_angle
```
Good point! I will add this.

kjbrooks commented:
@mperrin I have made the requested changes. Everything should be good to merge after the checks have passed.
